### PR TITLE
Add persistent content runtime logging

### PR DESCRIPTION
**Pull request by jdgleaver**
## Description

The current method for logging per-content playtime is very nice, but it has a few shortcomings. The most immediate of these is that the `content_runtime.lpl` playlist file is size limited - if you exceed the currently set history size, old records are lost. This behaviour is a requirement to prevent the performance impact of searching the `content_runtime.lpl` playlist from getting out of hand (it's O(n^2) at the moment, so making it too large will hurt junk-tier hardware), but it means we lack permanence.

This PR adds persistent runtime logging, by creating a separate runtime log file for each content+core combination:

- These are saved alongside existing content save files.

- They have the file extension `.lrtl` (Libretro RunTime Log)

- If the user has `Sort Saves In Folders` turned `ON`, the runtime log file path is: `<content_save_dir>/<content_name>.lrtl` (i.e. treated just like save files)

- If the user has `Sort Saves In Folders` turned `OFF`, the runtime log file path is: `<content_save_dir>/<content_name> - <core_name>.lrtl` (we have to differentiate the logs for each core like this, otherwise correctly pairing runtime with playlist entries will become impossible)

The runtime log files are plain text with the following example content:

```
3:27:39
2019-02-27 16:23:44
```

- 1st line: runtime hours:minutes:seconds

- 2nd line: last played timestamp (not currently used, but I have some plans for it later)

The files are deliberately kept as simple as possible (no metadata, etc.) to minimise read time. (See note in 'Reviewers' section!)

At present, the existing `content_runtime.lpl` playlist is updated synchronously with the runtime log files. Runtime logs are nominally used as the 'source' for runtime data, but for (temporary?) backwards compatibility any old `content_runtime.lpl` values are copied into the logs (where a `content_runtime.lpl` entry exists, but a runtime log file does not)

This should go a long way towards closing issue #3527. Still have the following work to do, however:

- Finalise `content_history.lpl`/`content_runtime.lpl` playlist handling, to increase efficiency of runtime display (and maybe enable it for all playlists?)

- Try to add 'playtime' and 'last played' entries for all content via the 'Information' submenu.

## Related Issues

#3527

## Reviewers

This is the first time I've added source files, so I would appreciate it if someone could check that I have done so correctly. I've added references to the new `runtime_file.c` src to `Makefile.common` and `griffin/griffin.c` - are these the only two required locations?

@twinaphex Could I please ask your opinion on the use of these log files in generating playlist sublabels?

Reading them is very fast - it has a far lower performance impact than I expected (although I deliberately kept them very light). On my old workstation (Athlon II X3 435) it takes a mean time of ~210ms to parse 10,000 runtime log files, which is almost silly. This being the case, I am strongly tempted to remove `content_runtime.lpl` entirely, and instead read the individual log files when populating playlist runtime values in:

https://github.com/libretro/RetroArch/blob/ebae7ebc794b9ee2a99f4647b8280cbbbb6cef57/menu/menu_displaylist.c#L1373-L1398

This would both:

- Get rid of the horrible double for loop that we currently have to use when parsing `content_history.lpl` (takes us back to ~O(n))

- Trivially allows us to also display runtime sublabels on all playlists, not just `content_history.lpl`

I would appreciate it if you could let me know what you think!

### CODE REVIEW

**Comment by twinaphex** [comment]
As long as performance doesn't suffer by doing this and as long as there are no big drawbacks to doing this, I see no issues with it whatsoever.

**Comment by hizzlekizzle** [comment]
I like this idea. It means playtimes are less likely to get carelessly deleted/reset, IMO.

**Comment by fr500** [comment]
why back to plain text now that we have a json parser?

**Comment by jdgleaver** [comment]
@fr500 Thanks for the feedback!

> why back to plain text now that we have a json parser?
> Ok seems you addressed this before..
> I dislike plain files to be honest because any new functioniality means to break backwards compatibility

I agree that plain text is uncomfortable, but these files are performance critical. The JSON parser adds a grotesque amount of bloat.

Hmm... Okay: let me run some tests to assess the real world performance impact of using JSON. If it's not so bad, then maybe we can go with it. But if it means we can't have runtime logging on 3DS/PSP/etc., then I'll have to insist on plain text...

> I dislike the use of the save directory for metadata, the save dir is used by the frontend or the core for saves.
> Also the file naming depending if it's being tracked by core or not seems weird.
>
> I'd use the playlist dir (say playlists/logs or whatever), after all it's more related to playlist functionality than anything else

On a conceptual level I could not disagree with you more. Runtime logs are user-generated data.

- User plays game -> save file is generated, recording progress.

- User plays game -> runtime log is generated, recording progress.

A runtime log is exactly the same kind of object as a save file, and belongs in the same place. Remember, logs are also generated when loading content via the file browser, or the command line. *They have nothing to do with playlists.*

**However....!**

From a purely practical perspective, what you say makes perfect sense. We currently have no way of actually *using* runtime log files without parsing an associated playlist, so it is entirely logical to group the two things together. Doing so also gets rid of the 'weird' file names (i.e. we are completely disentangled from the user's current save file config), and keeping the log files separate from saves makes it easier for the user to find/edit/transfer runtime logs by hand.

So I'm with you on this one. I'm going to change the log file path to the following:

`<playlist_dir>/logs/<core_name>/<content_name>.lrtl`


> and I'd use the same file for any case.
>
>playlist/logs/super mario kart (usa).whatever
>
>corename1
>3:27:39
>2019-02-27 16:23:44
>corename2
>3:27:39
>2019-02-27 16:23:44
>
>You could even have options for the aggregate play time or whatever like this. Without leveraging multiple files.

Not just no, but hell no.

Sorry, but I must stand firm here. The whole point of this is to break apart the runtime logs, for the sake of performance, scalability and logical consistency with other user generated content. If we group things together in the way you suggest then we're right back where we started.

Combining logs like this is as nonsensical to me as the idea of bundling different game saves into a single database object. It is wrong on every level. If that's the kind of show we're running, then count me out.

Apologies if that sounded a bit harsh, but I feel strongly about this! :smiley: 

---

@twinaphex Okay, I'm going to make a few changes/run some tests. Please don't merge until I report back :slightly_smiling_face: 

Also, since you are agreeable, I'm going to go ahead and retire `content_runtime.lpl`. Should I do this by continuing to work in this branch, or would you prefer a separate PR once this one is closed?

**Comment by jdgleaver** [comment]
@fr500 Well bugger me - I just tested this with JSON files and the read speed is only 7% slower than with plain text. Guess I had it figured all wrong... :slightly_smiling_face: 

Still have to implement the writer code, but this is going to be fine now. JSON all the way!

**Comment by twinaphex** [comment]
> Also, since you are agreeable, I'm going to go ahead and retire `content_runtime.lpl`. Should I do this by continuing to work in this branch, or would you prefer a separate PR once this one is closed?

What do you mean by retiring? It would be nice if we could have a solution that works both for JSON and the non-JSON playlists.

**Comment by hizzlekizzle** [comment]
What's the value of supporting the deprecated non-JSON playlists?

**Comment by jdgleaver** [comment]
@twinaphex Oh, no no no! I will not *remove* any existing functionality, only add.

What I mean is that we no longer need `content_runtime.lpl`, since we can read/write all runtime info from/to the individual log files directly (even with the new JSON format, runtime log file access is blisteringly fast, so no performance issues).

The result is that the playtime display will automatically work with *all* playlists (not just history), and the playlist format (old style or JSON) is completely irrelevant - we populate the runtime values *after* the playlist itself has been loaded, so everything is clean and distinct. (i.e. the current runtime entries in the playlist struct are purely for display purposes, and are never stored in playlist files on disk)

So basically, this PR + retiring `content_runtime.lpl` means that:

- Runtime logs are stored forever.

- Performance (relative to current, with sublabels enabled) is improved as playlist size increases.

- Play time can be shown everywhere.

- 'Last played' timestamp can be shown everywhere. (I think I may just end up adding this to playlist sublabels...)

It's a win-win all around  :slightly_smiling_face:

**Comment by jdgleaver** [comment]
Okay, just pushed the commit that adds JSON support.

As mentioned above, runtime log files now have the following path:

`<playlist_dir>/logs/<core_name>/<content_name>.lrtl`

...and they have the following content:

```
{
  "version": "1.0",
  "runtime": "1:28:27",
  "last_played": "2019-02-28 14:59:56"
}
```

On my old workstation, I can parse 10,000 log files in a mean time of ~230ms - not bad.

This is about as good as it gets, and I think this PR is ready to merge - unless you want me to finish the job of retiring `content_runtime.lpl` first? (or leave it for the next PR?)

**Comment by jdgleaver** [comment]
Sigh... Travis is now failing on the Apple builds due to various JSON* object redefinition errors.

Apple is crap.

Give me a couple of minutes to fix this...

**Comment by jdgleaver** [comment]
There, that seems to have fixed it. All good now.

**Comment by twinaphex** [comment]
OK then, let's merge it then.

**Comment by twinaphex** [comment]
@jdgleaver 
Some oversights here -

code that goes into libretro-common should be completely frontend-agnostic. Therefore, referencing RetroArch-only headers and files is a no-no. In this case, the issue is file_path_special.h.

**Comment by twinaphex** [comment]
Having just looked at it, I see no way how runtime_file.c belongs in libretro-common the way it's currently laid out.

It assumes core_info.h is there (another RetroArch-only file), it has all sorts of references to specific libretro cores. I'm going to move this file outside of libretro-common and move it into RetroArch. 

If you can later come up with something that makes sense from a libretro-common perspective, that is OK, but for now I need to make sure that libretro-common remains a cohesive ecosystem instead of all sorts of files referencing RetroArch-only files.

**Comment by twinaphex** [comment]
I moved runtime_file.c/.h to RetroArch for now.

I will also talk to @kivutar about making a new libretro site where we explain the design philosophies behind projects like libretro-common and RetroArch, since at this stage I think we have only ourselves to blame when it comes to understanding the design philosophies behind what goes in libretro-common and what doesn't. Beyond that , keep up the awesome work, nothing but love on our end.

**Comment by jdgleaver** [comment]
@twinaphex Thanks for sorting out the file paths. I only put runtime_file.c/.h in libretro-common because (a) it seemed to fit with config_file and archive_file and (b) I was hesitant to 'pollute' the main directory with new files, but yes - now you're explained it, I can see that was totally inappropriate :slightly_smiling_face: 

A website with more information for developers would be awesome - delving into a large codebase like RetroArch is pretty hard for outsiders, and it would be great to have some fixed guidelines laid out.

I'm really glad you appreciate my humble contributions. libretro/RetroArch is such a worthwhile project. I'm happy to make any little improvements where I can.

(I'll try to finalise the runtime log integration today - get playtime labels working with all playlists, and add the 'last played' timestamp. Will be nice to have this in place for the weekend!)

**Comment by fr500** [comment]
All good IMHO thanks for listening to feedback.

@Twinaphex that site already exists
docs.libretro.com
